tools/odbc/driver.cpp

```diff
@@ -126,65 +128,72 @@ SQLRETURN SQL_API SQLSetEnvAttr(SQLHENV environment_handle, SQLINTEGER attribute
 		case SQL_OV_ODBC2:
 			// TODO actually do something with this?
 			// auto version = (SQLINTEGER)(uintptr_t)value_ptr;
+			env->odbc_version = major_version;
 			return SQL_SUCCESS;
 		default:
-			env->error_messages.emplace_back("ODBC version not supported.");
-			return SQL_ERROR;
+			return duckdb::SetDiagnosticRecord(env, SQL_SUCCESS_WITH_INFO, "SQLSetEnvAttr",
+			                                   "ODBC version not supported: " + std::to_string(major_version),
+			                                   SQLStateType::ST_HY092, "");
 		}
 	}
-	case SQL_ATTR_CONNECTION_POOLING:
-		if (env) {
-			return SQL_ERROR;
-		}
-		switch ((SQLINTEGER)(intptr_t)value_ptr) {
+	case SQL_ATTR_CONNECTION_POOLING: {
+		auto pooling = static_cast<SQLINTEGER>(reinterpret_cast<intptr_t>(value_ptr));
+		switch (pooling) {
 		case SQL_CP_OFF:
 		case SQL_CP_ONE_PER_DRIVER:
 		case SQL_CP_ONE_PER_HENV:
+			env->connection_pooling = pooling;
 			return SQL_SUCCESS;
 		default:
 			return duckdb::SetDiagnosticRecord(env, SQL_SUCCESS_WITH_INFO, "SQLSetConnectAttr",
 			                                   "Connection pooling not supported: " + std::to_string(attribute),
-			                                   SQLStateType::INVALID_ATTR_OPTION_ID, "");
+			                                   SQLStateType::ST_HY092, "");
 		}
+	}
 	case SQL_ATTR_CP_MATCH:
-		env->error_messages.emplace_back("Optional feature not supported.");
-		return SQL_ERROR;
-	case SQL_ATTR_OUTPUT_NTS: /* SQLINTEGER */
-		switch (*(SQLINTEGER *)value_ptr) {
-		case SQL_TRUE:
+		return duckdb::SetDiagnosticRecord(env, SQL_SUCCESS_WITH_INFO, "SQLSetConnectAttr",
+		                                   "Optional feature not implemented.", SQLStateType::ST_HY092, "");
+	case SQL_ATTR_OUTPUT_NTS: /* SQLINTEGER */ {
+		auto output_nts = static_cast<SQLINTEGER>(reinterpret_cast<intptr_t>(value_ptr));
+		if (output_nts == SQL_TRUE) {
+			env->output_nts = SQL_TRUE;
 			return SQL_SUCCESS;
-		default:
-			env->error_messages.emplace_back("Optional feature not supported.");
-			return SQL_ERROR;
 		}
+		return duckdb::SetDiagnosticRecord(env, SQL_SUCCESS_WITH_INFO, "SQLSetConnectAttr",
+		                                   "Optional feature not implemented.  SQL_ATTR_OUTPUT_NTS must be SQL_TRUE",
+		                                   SQLStateType::ST_HY092, "");
+	}
 	default:
-		return SQL_ERROR;
+		return duckdb::SetDiagnosticRecord(env, SQL_SUCCESS_WITH_INFO, "SQLSetEnvAttr", "Invalid attribute value",
+		                                   SQLStateType::ST_HY024, "");
 	}
 }
 
 SQLRETURN SQL_API SQLGetEnvAttr(SQLHENV environment_handle, SQLINTEGER attribute, SQLPOINTER value_ptr,
                                 SQLINTEGER buffer_length, SQLINTEGER *string_length_ptr) {
+
 	if (value_ptr == nullptr) {
-		return SQL_ERROR;
+		return SQL_SUCCESS;
```

Comment:
I have the same curiosity here: how come this is `SQL_SUCCESS`?

Reply:
From the docs:
_If ValuePtr is NULL, StringLengthPtr will still return the total number of bytes (excluding the null-termination character for character data) available to return in the buffer pointed to by ValuePtr._

Except all the attributes are either `SQLUINTEGERS` or `SQLINTEGERS`, so the `StringLengthPtr` never gets set.  This is one of those standard ODBC function signatures that doesn't really make sense in this function.